### PR TITLE
[Bug]: Flower 1.32 runtime dependency install shadows the baked in-repo flip-utils in FL runs

### LINKED ISSUE #767: [Bug]: Flower 1.32 runtime dependency install shadows the baked in-repo flip-utils in FL runs
### Description

Flower **1.32**'s runtime dependency installation causes FL runs to silently execute the **PyPI release** of `flip-utils` instead of the **in-repo / baked** one.

When the SuperLink starts a ServerApp, Flower's superexec runs `uv sync` against the FAB's `pyproject.toml` into a fresh per-run environment and prepends it to `sys.path` (`flwr/supercore/superexec/dependency_installer.py`). The Flower app templates declare a PyPI-style dependency:

```toml
# fl-apps/flower/standard/pyproject.toml (and evaluation/)
dependencies = [
    "flip-utils>=0.1.8",
    ...
]
```

so uv resolves `flip-utils` from PyPI into that per-run env, which then shadows the package baked into the `flower-*` images — even though the baked one already satisfies the constraint (the per-run env is prepended to `sys.path`, so anything installed into it wins unconditionally).

### Impact

- **Local/branch FL testing is misleading.** You can rebuild the FL images from a branch (`make build-fl FL_BACKEND=flower`), confirm your `flip-utils` change is baked in, run an FL job — and the run still uses the PyPI release, so your change has no effect. This cost real debugging time while verifying #148: the baked `flip-utils` carried the new `x_label` code, but a Flower run used the PyPI package without it.
- **The platform's FL behaviour is pinned to whatever `flip-utils` is on PyPI at run time**, not to the code in this repo. A PyPI publish (or an old cached release) can change FL behaviour independently of the repo.
- Testing an **unpublished** `flip-utils` (a branch version not yet on PyPI) on-platform is impossible while the runtime install resolves from PyPI.

### Agreed fix (keep the Flower model, pin the source)

Keep Flower's runtime dependency installation enabled — per-run `uv sync` is the Flower deployment model and apps keep their own dependency lists — but steer the resolution via the template pyprojects' `[tool.uv]` tables:

- `fl-services/flower/fl-base/Dockerfile` keeps the in-repo flip-utils **source** at `/opt/flip-utils` (previously installed from `/tmp` and deleted).
- `fl-apps/flower/{standard,evaluation}/pyproject.toml` add `[tool.uv.sources] flip-utils = { path = "/opt/flip-utils" }` — uv never consults PyPI for a name with a source override, so every run builds flip-utils from the image-shipped source. Verified empirically: the path source wins even when PyPI has a newer satisfying release, and uv drops the version specifier for path-sourced packages (no fallback path exists).
- `torch`/`torchvision` are pinned to PyTorch's **cu128** index in the same tables (PyPI's default cu130 wheels need NVIDIA driver ≥580; hosts run 575.x) — mirrors the pin inside flip-utils' own pyproject, which uv does NOT honour for path dependencies (only the root project's `[tool.uv]` tables apply).
- SuperNodes opt in to ClientApp-side runtime install (`--allow-runtime-dependency-installation`, default-off in flwr 1.32) in the flower composes, so ServerApp and ClientApp resolve symmetrically.
- No dev bind-mounts — dev resolves exactly like prod: a flip-utils change (including unpublished-version testing) ships by rebuilding the FL images (`make build-fl FL_BACKEND=flower`) and restarting onto them. (An earlier revision bind-mounted the working tree over `/opt/flip-utils` in the dev composes; dropped in PR #772.)

`FLWR_DISABLE_RUNTIME_DEPENDENCY_INSTALLATION=1` (the workaround used during #745) is deliberately **not** adopted as the default — runtime installs stay on.

### Acceptance criteria

- [x] Flower FL runs (ServerApp and ClientApp) use the in-repo/baked `flip-utils`, not a runtime-installed PyPI copy, so branch changes to `flip-utils` take effect after an FL image rebuild.
- [x] Runtime dependency installation remains enabled on SuperLink and is enabled on SuperNodes across the flower composes.
- [x] Per-run torch resolves from the cu128 index (not PyPI's cu130 default).
- [x] Documented alongside the Flower FL setup (`fl-services/flower/README.md`, `create-flip-app-from-flower.rst`).

### Technical considerations

- Flower 1.32 upgrade: #664. The runtime installer machinery: `install_app_dependencies` runs `uv sync --no-install-project --no-install-package flwr --inexact` with `UV_PROJECT_ENVIRONMENT=<per-run env>`; only `flwr` itself is excluded from installation.
- NVFLARE is unaffected (no runtime dependency install — the image is the environment).
- The fl-apps templates are platform-controlled: `bundle_flower_application` never lets uploaded model files overwrite template files, and user files land under `app/`, so the root `pyproject.toml` (and its `[tool.uv]` tables) is always the platform's.
- **Known limitation:** apps submitted straight from `fl-tutorials/flower/` to the standalone stacks (`make -C fl-services/flower up` + `submit`) don't carry the pin and runtime-install their declared `flip-utils` from PyPI — documented in `fl-services/flower/README.md`.
- Prod trust hosts need outbound HTTPS to PyPI + `download.pytorch.org` for ClientApp runtime installs (noted in `compose_trust.production.flower.yml`).
- Rollout ordering: the template pin and the images that carry `/opt/flip-utils` ship together; a stale image with a new template fails the app launch loudly (uv errors on the missing path) rather than silently falling back to PyPI.

### PR DESCRIPTION
**Pull request by atriaybagur**
Closes #767.

## Problem

Flower ≥1.32 installs an app's declared dependencies at run time: at every app launch the superexec runs `uv sync` against the FAB's `pyproject.toml` into a fresh per-run environment and prepends it to `sys.path`. Our templates declare `flip-utils>=0.1.8`, so uv resolved **PyPI's** flip-utils into that env — silently shadowing the in-repo package baked into the `flower-*` images (same version number). Branch changes to flip-utils were no-ops on Flower runs, and platform behaviour was pinned to PyPI state at run time (#767, discovered while verifying #148; also bit #745/#757, which carried an env-var workaround).

## Design: keep the Flower model, pin the source

Runtime dependency installation **stays enabled** — per-run installs are Flower's deployment model and app dependency lists stay meaningful. What changes is where the platform-critical pieces resolve from, via `[tool.uv]` tables in the template pyprojects (which are platform-controlled: the bundler never lets uploaded files override them):

- **`fl-services/flower/fl-base/Dockerfile`** keeps the in-repo flip-utils source at **`/opt/flip-utils`** (previously installed from `/tmp` and deleted). The tree is **root-owned, read-only to `app`** (3ca204bb, from Copilot review): user-submitted apps run as the same `app` user in the same container, so a writable copy could be mutated to poison later runs' resolution. Its consumers only read it — flip-utils' hatchling backend builds in temp dirs — verified by rebuilding the images and re-running the `uv lock`/build checks below.
- **`fl-apps/flower/{standard,evaluation}/pyproject.toml`** add `[tool.uv.sources] flip-utils = { path = "/opt/flip-utils" }`. uv never consults the registry for a name with a source override (verified: the path source wins even when PyPI has a newer satisfying release, and uv drops the version specifier for path-sourced packages — no fallback path exists). The `flip-utils>=0.1.8` entry remains as standard metadata for non-uv consumers.
- **torch/torchvision pinned to PyTorch's cu128 index** in the same tables. This is load-bearing: PyPI's default torch is a **cu130** build requiring NVIDIA driver ≥580 (hosts run 575.x), and uv ignores a path-dependency's own `[tool.uv]` tables — only the root project's apply, so flip-utils' internal cu128 pin does not reach the per-run resolution.
- **SuperNodes opt in to ClientApp-side runtime install** (`--allow-runtime-dependency-installation`, default-off in flwr 1.32 unlike the SuperLink) across the four flower composes, so ServerApp and ClientApp resolve symmetrically.
- **No dev bind-mounts — dev resolves exactly like prod.** A flip-utils change ships the same way in every environment: rebuild the FL images (`make build-fl FL_BACKEND=flower` — the source lands at `/opt/flip-utils`) and restart onto them. (Earlier revisions of this branch bind-mounted the working tree over `/opt/flip-utils` in the dev composes for rebuild-free iteration; dropped in 860cf980 — the prod composes never mounted, and the pin never referenced the host.) Testing an unpublished flip-utils still needs no PyPI release and no version bump — just the rebuild.
- `FLWR_DISABLE_RUNTIME_DEPENDENCY_INSTALLATION=1` (the #745-era workaround) is deliberately **not** used; its compose entries were reverted from `feat/round-progress-card` (49350dfe).

**CI checker adjustment:** `check-tutorial-sync` paired the evaluation tutorial's pyproject with the template's byte-for-byte; those now diverge by design (the template carries the platform-only `[tool.uv]` tables, which would break workstation `uv sync` if copied into the tutorial), so the pyproject pair is dropped from the checker — the shared ServerApp/strategy code pairs remain guarded.

Docs: new "Runtime dependency installation" section in `fl-services/flower/README.md`, an on-platform dependency-resolution note in `create-flip-app-from-flower.rst`, a Flower egress row in the sys-admin firewall table (`admin-platform-support.rst`), and a CLAUDE.md line. A review pass (1440b34e) also tightened comment accuracy: the README staleness check inspects the changed file, the Dockerfile's "fails every app launch" is scoped to template-bundled apps, and flip-utils' own torch comment now correctly says PyPI's 2.11.x default is cu130.

## Verification

Static: `uv lock` **inside the rebuilt image** against the standard template resolves the full tree (111 packages) with `flip-utils` → `source = { directory = "/opt/flip-utils" }` and `torch` → `registry = "https://download.pytorch.org/whl/cu128"`. Empirical uv-semantics check: a path source wins even when PyPI has a newer satisfying release, and uv drops the version specifier for path-sourced packages — so there is no PyPI fallback path.

Live: rebuilt the flower images from this branch (`make build-fl FL_BACKEND=flower`), switched the dev stack onto them, added a **temporary marker print to working-tree flip-utils only** (never rebuilt — deliberately, to exercise the dev bind-mount), and ran the full e2e smoke (`make e2e_smoke FL_BACKEND=flower`, xray classification, both trusts). During training on net-2:

- **ClientApps (both trusts)** logged the per-run install `flip-utils==0.3.0 (from file:///opt/flip-utils)` among the resolved tree, then printed the marker at import time from the per-run env:
  `FLIP767-VERIFY unpublished flip-utils 0.3.0 imported from /app/.flwr/runtime-envs/10515128689134631965/lib/python3.12/site-packages/flip/__init__.py`
- **ServerApp** (superexec output doesn't reach the container stdout, so verified on the filesystem): the per-run env's `flip_utils-0.3.0.dist-info/direct_url.json` reads `{"url":"file:///opt/flip-utils"}`, and the installed copy contains the marker line — proving the bind-mounted **working tree** (not the image layer, which predates the marker) was the build source. The rebuild-free dev loop works on both sides.
- **torch resolved as `2.11.0+cu128`** in both the ServerApp and ClientApp per-run envs — the index pin held; PyPI's default cu130 build (driver ≥580) never entered the environment.
- e2e completed the full lifecycle: cohort → approval → image pull → training on both trusts → results uploaded + downloaded (`🎉 Smoke passed`, `final_status = RESULTS_UPLOADED`).

One unrelated incident during the run, for the record: flip-api's DB pool wedged mid-training (all 15 connections `idle in transaction`, parked after the trust-lookup SELECT — a pre-existing session leak under the metrics burst; nothing on this branch touches flip-api). One flip-api restart cleared it and the run completed. A separate bug report is drafted for it.

The marker was removed before committing.

**Follow-up — mounts dropped (860cf980):** the live run above exercised the then-present dev bind-mount (the marker lived only in the working tree). After removing the mounts, the resolution path was re-verified with **no volumes attached**: `uv lock` against the standard template inside the unmounted branch-built supernode image resolves `flip-utils` → `source = { directory = "/opt/flip-utils" }` (the image-baked copy) and torch → `2.11.0+cu128` — the same resolution the superexec performs per run. All four flower composes (hub dev base+overlay, trust dev base+overlay, standalone dev, standalone secure) render cleanly after the removal.

Observed per-round behaviour worth knowing: the `flwr-clientapp` process exits after each message and its exit handler removes the per-run env, so **every round re-runs the install** — the first round downloads the tree (multi-GB, minutes), subsequent rounds rebuild the env from uv's warm cache via hardlinks (seconds). The pin held identically on every round.

## Rollout notes

- **Merge order:** #767 before #757 — the RoundProgress branch's round telemetry needs the in-repo `flip` package at run time.
- **Prod trust hosts need outbound HTTPS to PyPI + `download.pytorch.org`** for ClientApp runtime installs (noted in `compose_trust.production.flower.yml` and added to the sys-admin firewall-rules table). Previously trusts only needed outbound HTTPS to the hub.
- Images and templates ship together: a stale image (no `/opt/flip-utils`) with a new template fails the app launch loudly (uv errors on the missing path) rather than silently falling back to PyPI.
- **Known limitation:** apps submitted straight from `fl-tutorials/flower/` to the standalone stacks don't carry the pin and runtime-install their declared flip-utils from PyPI (plain Flower semantics) — documented in the README with the ad-hoc pin workaround.

## Acceptance Criteria

*Imported from issue #767*

- [x] Flower FL runs (ServerApp and ClientApp) use the in-repo/baked `flip-utils`, not a runtime-installed PyPI copy, so branch changes to `flip-utils` take effect after an FL image rebuild.
- [x] Runtime dependency installation remains enabled on SuperLink and is enabled on SuperNodes across the flower composes.
- [x] Per-run torch resolves from the cu128 index (not PyPI's cu130 default).
- [x] Documented alongside the Flower FL setup (`fl-services/flower/README.md`, `create-flip-app-from-flower.rst`).

### CODE REVIEW

**Comment by atriaybagur** [comment]
Ready for review 🎉

- **CI**: all 31 checks green on `1440b34e` (the main-origin guard skipping is expected for a develop-targeted PR); mergeable against `develop`.
- **Design update since the original description**: the dev bind-mounts over `/opt/flip-utils` were dropped (860cf980) — dev now resolves exactly like prod, and a flip-utils change ships via image rebuild everywhere. Verified with no mounts attached: `uv lock` inside the rebuilt image resolves `flip-utils` from `directory=/opt/flip-utils` and torch from the cu128 index. (The earlier live e2e on both trusts had already exercised the full pipeline.)
- **Review pass applied** (1440b34e): Flower egress row added to the sys-admin firewall table (trust hosts now need outbound 443 to PyPI + `download.pytorch.org`), README staleness-check command fixed, `create-flip-app-from-flower.rst` 'pre-installed' note aligned with the per-run install mechanism, Dockerfile comment scoped to template-bundled apps, and flip-utils' torch comment corrected (PyPI 2.11.x default is cu130).
- **Acceptance criteria**: all four ticked in the PR body and issue #767.
- Copilot review requested.

File: fl-services/flower/fl-base/Dockerfile [hunk 1]
```diff
@@ -27,18 +27,24 @@ WORKDIR /app
...
-RUN python -m pip install -U --no-cache-dir "/tmp/flip-utils[full]" \
-      --extra-index-url https://download.pytorch.org/whl/cu128 \
-  && rm -rf /tmp/flip-utils
+# to root ownership — the runtime installer builds flip-utils from this dir as
+# `app`.
+COPY --chown=app:app flip-utils/pyproject.toml flip-utils/README.md /opt/flip-utils/
+COPY --chown=app:app flip-utils/docs/ /opt/flip-utils/docs/
+COPY --chown=app:app flip-utils/flip/ /opt/flip-utils/flip/
+RUN python -m pip install -U --no-cache-dir "/opt/flip-utils[full]" \
+      --extra-index-url https://download.pytorch.org/whl/cu128
```
[the remark this answers is not in the document]
Comment (atriaybagur) [reply fl-services/flower/fl-base/Dockerfile:50]: Fixed in 3ca204bb: the flip-utils COPYs now keep COPY's default root ownership, so `/opt/flip-utils` is read-only to the `app` user that both the flwr processes and user-submitted apps run as. Verified in the rebuilt image: create/append/overwrite as `app` all denied; the build-time `pip install "/opt/flip-utils[full]"` still succeeds and the per-run resolution is unchanged (`uv lock` against the standard template resolves flip-utils from `directory = "/opt/flip-utils"`, torch `2.11.0+cu128`; a cold-cache `uv pip install /opt/flip-utils` builds cleanly) — flip-utils's hatchling backend builds in temp dirs and never writes to the source tree.

Scope note: this closes the flagged persistence path but does not create a privilege boundary inside the container — the base venv's site-packages and uv's cache under `/app` are necessarily app-writable, so in-container code can still poison those. The primary control on user-submitted code remains the pre-run review/hashing gate; this is defense-in-depth for the pinned source specifically.

**Review by garciadias — APPROVED** [review]
Solid fix for Flower 1.32 runtime dependency install shadowing the baked flip-utils (FLIP#767). Key decisions correct:

- /opt/flip-utils root-owned, read-only to app user — prevents app code from poisoning the source tree while still allowing per-run uv builds (hatchling builds in temp dirs)
- [tool.uv.sources] pin in template pyprojects, not tutorials — correct separation: tutorials drive workstation venvs where /opt/flip-utils doesn't exist
- SuperNodes opt in via --allow-runtime-dependency-installation (explicit; default-off in flwr 1.32)
- Torch cu128 index pin mirrors fl-base Dockerfile but in the right place (uv honors only root project's [tool.uv])
- Tutorial sync check intentionally excludes pyproject.toml (documented in check_tutorial_sync.sh)
- Trust compose files (dev + prod) updated with the flag

Docs are thorough: flower README section on runtime dependency install, platform-support.rst firewall note, create-flip-app-from-flower.rst resolver explanation.
